src/Collateral.sol

```diff
@@ -0,0 +1,202 @@
+// Copyright 2023-, Semiotic AI, Inc.
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import {TAPVerifier} from "./TAPVerifier.sol";
+import {AllocationIDTracker} from "./AllocationIDTracker.sol";
+
+import "forge-std/console.sol";
+
+/**
+ * @title Collateral
+ * @dev This contract allows `senders` to deposit collateral for specific `receivers`,
+ *      which can later be redeemed using Receipt Aggregate Vouchers (`RAV`) signed
+ *      by an authorized `signer`. `Senders` can deposit collateral for `receivers`,
+ *      authorize `signers` to create signed `RAVs`, and withdraw collateral after a
+ *      set `thawingPeriod` number of blocks. `Receivers` can redeem signed `RAVs` to
+ *      claim collateral.
+ * @notice This contract uses the `TAPVerifier` contract for recovering signer addresses
+ *         from `RAVs`.
+ */
+contract Collateral {
+    struct CollateralAccount {
+        uint256 balance; // Total collateral balance for a sender-receiver pair
+        uint256 amountThawing; // Amount of collateral currently being thawed
+        uint256 thawEnd; // Block number at which thawing period ends
+    }
+
+    // Stores how much collateral each sender has deposited for each receiver, as well as thawing information
+    mapping(address => mapping(address => CollateralAccount)) private collateralAccounts;
+    // Map of authorized signers to which sender they are authorized to sign for
+    // key: signer address, value: sender address
+    mapping(address => address) private authorizedSigners;
```

Comment:
If we upgrade solidity version to 0.8.18 we could used named parameters in mapping types which I think are great to improve code readability.

Note that this applies to any mapping but specially this one since it required an extra comment to explain what key and value are.

https://github.com/ethereum/solidity/blob/develop/Changelog.md#0818-2023-02-01

Reply:
I am moving this to an issue to be addressed on a separate PR (#7 )